File: java/0004_Median_of_Two_Sorted_Arrays.java
class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        int fullLength = nums1.length + nums2.length;
        int midCount = 2 - (fullLength % 2);
        int cnt = fullLength >> 1;
        double mid = 0;

        int i = 0;
        int j = 0;

        while (cnt >= 0) {
            int tmpMid;

            if (i < nums1.length && j < nums2.length) {
                tmpMid = (nums1[i] > nums2[j]) ? nums2[j++] : nums1[i++];
            } else {
                tmpMid = (i >= nums1.length) ? nums2[j++] : nums1[i++];
            }

            if (cnt < midCount) {
                mid += tmpMid;
            }
            cnt--;
        }

        return mid / midCount;
    }
}

File: kotlin/0004_Median_of_Two_Sorted_Arrays.kt
class Solution {
    fun findMedianSortedArrays(nums1: IntArray, nums2: IntArray): Double {
        val fullLength = nums1.size + nums2.size
        val midCount = 2 - fullLength % 2
        var cnt = fullLength shr 1

        var mid = 0.0
        var i = 0
        var j = 0

        while (cnt >= 0) {
            val tmpMid: Int =
                if (i < nums1.size && j < nums2.size) {
                    if (nums1[i] > nums2[j]) nums2[j++] else nums1[i++]
                } else {
                    if (i >= nums1.size) nums2[j++] else nums1[i++]
                }

            if (cnt < midCount) {
                mid += tmpMid.toDouble()
            }
            cnt--
        }
        return mid / midCount
    }
}